File: GDCarFinder/GDCarFinder/UI/ViewController/GDMapViewController.m
//
//  GDMapViewController.m
//  GDCarFinder
//
//  Created by Guglielmo Deletis on 29/05/22.
//

#import "GDMapViewController.h"
#import "GDCarFinder-Swift.h"

@interface GDMapViewController ()
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (strong, nonatomic) id<GDMapHandler> mapHandler;

@end

@implementation GDMapViewController
- (id)init {
    [self doesNotRecognizeSelector:_cmd];
    return self;
}

- (id)initWithMapHandler:(id<GDMapHandler>)handler {
    self = [super initWithNibName:@"GDMapViewController" bundle:nil];
    self.mapHandler = handler;
    return self;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    [self doesNotRecognizeSelector:_cmd];
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    [self.mapHandler viewDidLoadWithView:self.mapView];
}

#pragma mark MKMapViewDelegate
- (void)mapView:(MKMapView *)mapView regionDidChangeAnimated:(BOOL)animated {
    [self.mapHandler mapViewWithRegionDidChange:mapView];
}

- (void)mapViewWillStartLoadingMap:(MKMapView *)mapView {
    
}

- (void)mapViewDidFinishLoadingMap:(MKMapView *)mapView {
    [self.mapHandler mapViewWithDidFinishLoadingMap:mapView];
}

- (void)mapViewDidFailLoadingMap:(MKMapView *)mapView withError:(NSError *)error {
    [self.mapHandler mapViewWithDidFinishLoadingMap:mapView];
}

@end

File: GDCarFinder/GDCarFinder/UI/ViewController/GDMapViewController.swift
//
//  GDMapViewController.swift
//  GDCarFinder
//
//  Created by Guglielmo Deletis on 29/05/22.
//

import Foundation
import UIKit
import MapKit

class GDMapViewController: GDBaseViewController {
    @IBOutlet weak var mapView: MKMapView!
    @IBOutlet weak var loadingView: UIView!
    private var mapHandler: GDMapHandler
    
    init(loader:GDLoader, mapHandler: GDMapHandler) {
        self.mapHandler = mapHandler
        super.init(loader: loader, nibName: "GDMapViewController", bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.mapHandler.viewDidLoad(view: self.mapView)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.loader.delegate = self
        if !self.loader.isLoading {
            if !self.loader.hasFailed {
                self.loadAnnotations()
            } else {
                self.loader.load(urlString: GDConst.defaultVehicleListURLString, handler: GDOperationQueueManager.instance)
            }
        }
    }
    
    private func loadAnnotations() {
        self.mapHandler.loadAnnotations(inMapView: self.mapView)
    }
}

extension GDMapViewController: MKMapViewDelegate {
    //MARK: MKMapViewDelegate
    func mapViewDidChangeVisibleRegion(_ mapView: MKMapView) {
        self.mapHandler.mapView(regionDidChange: mapView)
    }
    
    func mapViewDidFinishLoadingMap(_ mapView: MKMapView) {
        self.mapHandler.mapView(didFinishLoadingMap: mapView)
    }
    
    func mapViewDidFailLoadingMap(_ mapView: MKMapView, withError error: Error) {
        self.mapHandler.mapView(didFinishLoadingMap: mapView)
    }
}

//MARK: GDLoaderDelegate
extension GDMapViewController: GDLoaderDelegate {
    func loaderDidStart(_ loader: GDLoader) {
        DispatchQueue.main.async {
            self.loadingView.isHidden = false
        }
    }
    
    func loaderDidLoad(_ loader: GDLoader, data: [Data]?) {
        if let d = data?[0] {
            self.mapHandler.listFromData(d)
        }
        DispatchQueue.main.async {
            self.loadingView.isHidden = true
            self.loadAnnotations()
        }
    }
    
    func loaderFailed(_ loader: GDLoader, error: Error) {
        DispatchQueue.main.async {
            self.loadingView.isHidden = true
            self.showError(error)
        }
    }
    
    func loaderCancelled(_ loader: GDLoader) {
        DispatchQueue.main.async {
            self.loadingView.isHidden = true
        }
    }
}
